### DISCUSSION TITLE
Additive synthesis efficient sine?

### DISCUSSION THREAD

**Discussion by elanhickler** [RobinSchmidt/RS-MET#331] [Q&A]
Robin, do you have an efficient sine algorithm for additive synthesis?

**Comment by elanhickler** [comment]
some efficient algorithms here: https://www.kvraudio.com/forum/viewtopic.php?t=474872

**Comment by RobinSchmidt** [comment]
i have recently found a nice recursion to efficiently compute sines with polynomial envelops for instantaneous phase and amplitude, see here

https://www.kvraudio.com/forum/viewtopic.php?f=33&t=567563&p=8164942#p8164945

here is the math:

https://github.com/RobinSchmidt/RS-MET/blob/master/Notes/ExpOfPolyRecursion.txt

and here is the code (class rsSineSweepIterator):

https://github.com/RobinSchmidt/RS-MET/blob/master/Libraries/RobsJuceModules/rapt/Math/Functions/FunctionIterators.h
https://github.com/RobinSchmidt/RS-MET/blob/master/Libraries/RobsJuceModules/rapt/Math/Functions/FunctionIterators.cpp


the well known recursions are only for producing sines with fixed frequency and amplitude. for an additive engine, such enveloping capability could certainly be useful

**Comment by RobinSchmidt** [comment]
i've started writing some code based on this idea:

https://github.com/RobinSchmidt/RS-MET/blob/master/Libraries/RobsJuceModules/rs_testing/Prototypes/AdditiveSynthEngine.h
https://github.com/RobinSchmidt/RS-MET/blob/master/Libraries/RobsJuceModules/rs_testing/Prototypes/AdditiveSynthEngine.cpp
